### PR TITLE
cleanup static_cast of AutogradMeta

### PR DESCRIPTION
**Pull request by albanD**
The goal is to reduce the spread of static casts in the autograd code as per the comment in https://github.com/pytorch/pytorch/pull/49097#discussion_r543695091
I wasn't sure how to use a virtual method here but a simple method in impl clean it up quite nicely.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54103 cleanup static_cast of AutogradMeta**
* #54102 properly make AutogradMeta/DifferentiableViewMeta attributes internal
* #54101 Remove legacy from optional-related function names
* #54100 Add size check for forward grads
* #54099 make internal forwardAD methods on at::Tensor internal

Differential Revision: [D27117840](https://our.internmc.facebook.com/intern/diff/D27117840)

### CODE REVIEW

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 1]
```diff
@@ -178,19 +177,20 @@ inline Tensor as_view(const Tensor & base, const Tensor & tensor, bool is_bw_dif
...
     // Check if base is a forward differentiable view
     auto base_meta = torch::autograd::impl::get_autograd_meta(base);
-    auto is_view = base_meta && base_meta->is_view_;
-    if (is_view && static_cast<DifferentiableViewMeta*>(base_meta)->has_fw_view()) {
-      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(base_meta);
-      const auto& base_fw_info = diff_view_meta->get_forward_view();
-      new_fw_info = base_fw_info.chain(base, tensor, view_func);
+    if (base_meta && base_meta->is_view_) {
+      auto diff_view_meta = torch::autograd::impl::get_view_autograd_meta(base);
+      if (diff_view_meta->has_fw_view()) {
```
Comment (soulitzer) [inline torch/csrc/autograd/VariableTypeUtils.h:182]: Is it intended that the behavior here changed? If this if condition is false, then we will no longer go into the else.

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 2]
```diff
@@ -225,22 +225,23 @@ inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor>& ten
...
-      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(base_meta);
-      const auto& base_fw_info = diff_view_meta->get_forward_view();
-      TORCH_INTERNAL_ASSERT(creation_meta == CreationMeta::MULTI_OUTPUT_NODE || creation_meta == CreationMeta::MULTI_OUTPUT_SAFE,
-                            "Functions that result multiple view must have a creation meta reflecting this behavior.");
-      // It is ok to create a ViewInfo where only the base is correct in this case as inplace operations on such views are
-      // not allowed
-      new_fw_info = ViewInfo(base_fw_info.base_, /* view_func */ nullptr);
+    if (base_meta && base_meta->is_view_) {
+      auto diff_view_meta = torch::autograd::impl::get_view_autograd_meta(base);
+      if (diff_view_meta->has_fw_view()) {
```
Comment (soulitzer) [inline torch/csrc/autograd/VariableTypeUtils.h:230]: Same for here?

File: torch/csrc/autograd/variable.cpp [hunk 3]
```diff
@@ -322,6 +320,12 @@ namespace impl {
     return static_cast<AutogradMeta*>(self.unsafeGetTensorImpl()->autograd_meta());
   }
 
+  DifferentiableViewMeta* get_view_autograd_meta(const Variable& self) {
```
Comment (soulitzer) [inline torch/csrc/autograd/variable.cpp:323]: nit: Would we want to use `Tensor` here for our variables names here? or do we keep `Variable` for consistency.

Comment (albanD) [reply torch/csrc/autograd/variable.cpp:323]: Yes we have a `using Variable = at::Tensor` at the top of the header.
We should remove that and rename all of them I think yes.

File: torch/csrc/autograd/VariableTypeUtils.h (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/VariableTypeUtils.h:182]: Good catch.
I wanted to reduce the amount of time we access the autograd meta but that's not really possible I guess.
I'll revert back to the old logic

**Comment by ezyang** [comment]
This seems OK, but given that all the call sites are testing for view-ness before calling the cast, it seems like it might just be better to make this function return a nullptr/ullopt when it's not castable, and then force callers to test the return output.

**Comment by albanD** [comment]
@ezyang I updated following your recommendation. Does that look better? It does make the code simpler but it is a less clear that we use this to check if a Tensor is a view.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
One way to make it clearer is to jiggle the naming; e.g., `maybe_view_XXX` or `try_view_XXX`

**Pull request closed without merging**